Add startup check to verify existing DB data matches the algod network.

shiqizng opened the pull request:
<!--
Thanks for submitting a pull request! We appreciate the time and effort you spent to get this far.

If you haven't already, please make sure that you've reviewed the CONTRIBUTING guide:
https://github.com/algorand/go-algorand/blob/master/CONTRIBUTING.md#code-guidelines
-->

## Summary
Write and check genesis hash during db load genesis. 

## Test Plan
Add a test to postgres_integration_test

tolikzinovyev commented on idb/postgres/postgres.go:
```diff
@@ -397,9 +397,29 @@ func (db *IndexerDb) AddBlock(block *bookkeeping.Block) error {
 func (db *IndexerDb) LoadGenesis(genesis bookkeeping.Genesis) error {
 	f := func(tx pgx.Tx) error {
+		// check networkdID
+		network, err := db.getMetastate(context.Background(), nil, schema.NetworkMetaStateKey)
```
suggest making `getNetworkState()` and `setNetworkState()` similar to `?etImportState()` and `?etMigrationState()`.

tolikzinovyev commented on idb/postgres/postgres.go:
```diff
@@ -397,9 +397,29 @@ func (db *IndexerDb) AddBlock(block *bookkeeping.Block) error {
+				return fmt.Errorf("LoadGenesis() err: %w", err)
+			}
+			if networkState.NetworkID != string(genesis.Network) {
+				return fmt.Errorf("LoadGenesis() err: NetworkID does not match")
```
Styling: "err: " means `err` is a variable. Just write `return fmt.Errorf("LoadGenesis() NetworkID does not match")`.

tolikzinovyev commented on idb/postgres/postgres_integration_test.go:
```diff
@@ -1876,3 +1876,29 @@ func TestTransactionsTxnAhead(t *testing.T) {
 }
+
+func TestNetworkIDCheckAtDBStartup(t *testing.T) {
+
```
empty line

tolikzinovyev reviewed (commented):
Is network ID enough? There might be some other parameters. Why not use the genesis hash?

shiqizng commented:
> Is network ID enough? There might be some other parameters. Why not use the genesis hash?

genesis hash isn't in bookkeeping.Genesis struct and so the hash is not available in LoadGenesis(genesis bookkeeping.Genesis).

tolikzinovyev commented:
> > Is network ID enough? There might be some other parameters. Why not use the genesis hash?
> 
> genesis hash isn't in bookkeeping.Genesis struct and so the hash is not available in LoadGenesis(genesis bookkeeping.Genesis). Maybe it's better to use Genesis.ID()?
> 
> ```
> // ID is the effective Genesis identifier - the combination
> // of the network and the ledger schema version
> func (genesis Genesis) ID() string {
> 	return string(genesis.Network) + "-" + genesis.SchemaID
> }
> ```

Seems like there is infrastructure for hashing `bookkeeping.Genesis`. https://github.com/algorand/go-algorand/blob/master/data/bookkeeping/genesis.go#L121 I just don't know how to invoke it.

shiqizng commented:
> > > Is network ID enough? There might be some other parameters. Why not use the genesis hash?
> > 
> > 
> > genesis hash isn't in bookkeeping.Genesis struct and so the hash is not available in LoadGenesis(genesis bookkeeping.Genesis). Maybe it's better to use Genesis.ID()?
> > ```
> > // ID is the effective Genesis identifier - the combination
> > // of the network and the ledger schema version
> > func (genesis Genesis) ID() string {
> > 	return string(genesis.Network) + "-" + genesis.SchemaID
> > }
> > ```
> 
> Seems like there is infrastructure for hashing `bookkeeping.Genesis`. https://github.com/algorand/go-algorand/blob/master/data/bookkeeping/genesis.go#L121 I just don't know how to invoke it.

I changed the check to use genesis hash. I removed network ID check since the hash includes the ID also.

tolikzinovyev commented on idb/postgres/internal/encoding/encoding_test.go:
```diff
@@ -4,11 +4,13 @@ import (
 	"github.com/algorand/go-algorand/data/transactions"
 	"github.com/algorand/go-algorand/ledger/ledgercore"
 	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/indexer/idb/postgres/internal/types"
```
this should move below

tolikzinovyev commented on idb/postgres/postgres.go:
```diff
@@ -397,9 +398,26 @@ func (db *IndexerDb) AddBlock(block *bookkeeping.Block) error {
 func (db *IndexerDb) LoadGenesis(genesis bookkeeping.Genesis) error {
 	f := func(tx pgx.Tx) error {
+		// check genesis hash
+		network, err := db.getNetworkState(context.Background(), nil)
```
this should use `tx`

tolikzinovyev commented on idb/postgres/postgres.go:
```diff
@@ -397,9 +398,26 @@ func (db *IndexerDb) AddBlock(block *bookkeeping.Block) error {
+			networkState := types.NetworkState{
+				GenesisHash: crypto.HashObj(genesis),
+			}
+			err = db.setNetworkState(nil, &networkState)
```
this should use `tx`

tolikzinovyev reviewed (approved):
LGTM
